gather

```diff
@@ -49,41 +39,86 @@ def run(options=None):
+    for source in sources:
+        extra = {}
 
-        # Unless --include-parents is specified, exclude them.
-        if not include_parents:
-            if (domain == base) or (domain == "www.%s" % base):
+        try:
+            gatherer = importlib.import_module("gatherers.%s" % source)
+        except ImportError:
+            # If it's not a registered module, allow it to be "hot registered"
+            # as long as the user gave us a flag with that name that can be
+            # used as the --url option to the URL module.
+            if options.get(source):
+                gatherer = importlib.import_module("gatherers.url")
+                extra['name'] = source
+            else:
+                exc_type, exc_value, exc_traceback = sys.exc_info()
+                logging.error("[%s] Gatherer not found, or had an error during loading.\n\tERROR: %s\n\t%s" % (source, exc_type, exc_value))
+                exit(1)
+
+        # Iterate over each hostname.
+        for domain in gatherer.gather(suffix, options, extra):
+
+            # Always apply the suffix to returned names.
+            if not suffix_pattern.search(domain):
                 continue
 
-        # Apply --parent domain whitelist, if present.
-        if parents:
-            if base not in parents:
-                continue
+            base = utils.base_domain_for(domain)
+
+            # Unless --include-parents is specified, exclude them.
+            if not include_parents:
+                if (domain == base) or (domain == "www.%s" % base):
+                    continue
+
+            # Apply --parent domain whitelist, if present.
+            if parents:
+                if base not in parents:
+                    continue
 
-        # Use hostname cache to de-dupe, if seen before.
-        if domain not in hostnames_cache:
+            # Use hostname cache to de-dupe, if seen before.
+            if domain not in hostnames_cache:
+                hostnames_cache[domain] = [source]
+            elif source not in hostnames_cache[domain]:
+                hostnames_cache[domain] += [source]
 
-            # Finally, write out the hostname, and cache it.
-            gathered_writer.writerow([domain, base])
-            hostnames_cache[domain] = None
+    # Now that we've gone through all sources and logged when each
+    # domain appears in each one, go through cache and write
+    # all of them to disk.
 
+    # Assemble headers.
+    headers = ["Domain", "Base Domain"]
+    # Add headers dynamically for each source.
+    headers += sources
+
+    # Open CSV file.
+    gathered_filename = "%s/%s.csv" % (utils.results_dir(), filename)
+    gathered_file = open(gathered_filename, 'w', newline='')
+    gathered_writer = csv.writer(gathered_file)
+    gathered_writer.writerow(headers)
+
+    # Write each hostname to disk, with all discovered sources.
+    hostnames = list(hostnames_cache.keys())
+    hostnames.sort()
```

Comment:
Hostnames are always all sorted in-memory before being output -- otherwise the resulting output would be too unpredictable. But this does mean that super-large gathering jobs might become unwieldy.